das/redis_storage_service.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"context"
+	"crypto/subtle"
+	"encoding/base32"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type RedisConfig struct {
+	RedisUrl   string        `koanf:"redis-url"`
+	Expiration time.Duration `koanf:"redis-expiration"`
+	KeyConfig  string        `koanf:"redis-key-config"`
+}
+
+var DefaultRedisConfig = RedisConfig{
+	RedisUrl:   "",
+	Expiration: time.Hour,
+	KeyConfig:  "",
+}
+
+func RedisConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".redis-url", DefaultRedisConfig.RedisUrl, "Redis url")
+	f.Duration(prefix+".redis-expiration", DefaultRedisConfig.Expiration, "Redis expiration")
+	f.String(prefix+".redis-key-config", DefaultRedisConfig.KeyConfig, "Redis key config")
+}
+
+type RedisStorageService struct {
+	baseStorageService StorageService
+	redisConfig        RedisConfig
+	signingKey         common.Hash
+	client             redis.UniversalClient
+}
+
+var keyIsHexRegex = regexp.MustCompile("^(0x)?[a-fA-F0-9]{64}$")
+
+func NewRedisStorageService(redisConfig RedisConfig, baseStorageService StorageService) (StorageService, error) {
+	redisOptions, err := redis.ParseURL(redisConfig.RedisUrl)
+	if err != nil {
+		return nil, err
+	}
+	keyIsHex := keyIsHexRegex.Match([]byte(redisConfig.KeyConfig))
+	if keyIsHex {
+		return nil, errors.New("signing key file contents are not 32 bytes of hex")
+	}
+	signingKey := common.HexToHash(redisConfig.KeyConfig)
+	return &RedisStorageService{
+		baseStorageService: baseStorageService,
+		redisConfig:        redisConfig,
+		signingKey:         signingKey,
+		client:             redis.NewClient(redisOptions),
+	}, nil
+}
+
+func (rs *RedisStorageService) verifyMessageSignature(data []byte) ([]byte, error) {
+	if len(data) < 32 {
+		return nil, errors.New("data is too short to contain message signature")
+	}
+	message := data[32:]
+	var haveHmac common.Hash
+	copy(haveHmac[:], data[:32])
+	expectHmac := crypto.Keccak256Hash(rs.signingKey[:], message)
+	if subtle.ConstantTimeCompare(expectHmac[:], haveHmac[:]) == 1 {
```

Comment:
The customary golang style would invert the sense of this `if` statement, so that early termination happened in the error case, and the end-of-function return statement is the non-error case.